lib/ide/src/view/text_editor.rs

```diff
@@ -2,36 +2,70 @@
 
 use crate::prelude::*;
 
-use super::temporary_panel::TemporaryPanel;
-use super::temporary_panel::TemporaryPadding;
+use crate::view::notification::NotificationService;
+use crate::view::temporary_panel::TemporaryPadding;
+use crate::view::temporary_panel::TemporaryPanel;
 
 use basegl::display::object::DisplayObjectOps;
 use basegl::display::shape::text::glyph::font::FontRegistry;
 use basegl::display::shape::text::text_field::TextField;
 use basegl::display::shape::text::text_field::TextFieldProperties;
 use basegl::display::world::*;
-
+use enso_frp::io::Key;
+use enso_frp::io::KeyboardActions;
+use enso_frp::io::KeyMask;
 use nalgebra::Vector2;
 use nalgebra::zero;
 
 
+
 // ==================
 // === TextEditor ===
 // ==================
 
+shared! { TextEditor
+
 /// TextEditor allows us to edit text files or Enso Modules. Extensible code highlighting is
 /// planned to be implemented for it.
-#[derive(Clone,Debug)]
-pub struct TextEditor {
-    text_field : TextField,
-    padding    : TemporaryPadding,
-    position   : Vector2<f32>,
-    size       : Vector2<f32>
+#[derive(Debug)]
+pub struct TextEditorData {
+    text_field           : TextField,
+    padding              : TemporaryPadding,
+    position             : Vector2<f32>,
+    size                 : Vector2<f32>,
+    controller           : controller::text::Handle,
+    notification_service : NotificationService
```

Comment:
This is important. `NotificationService` duplicates the logic of logger. Instead of using a notification service explicitly, "warning", and "error" logs should be duplicated to it.